storage/SplitRowStoreTupleStorageSubBlock.cpp

```diff
@@ -194,379 +257,125 @@ TupleStorageSubBlock::InsertResult SplitRowStoreTupleStorageSubBlock::insertTupl
-                                            relation_.numNullableAttributes());
-          tuple_null_bitmap.clear();
-          char *fixed_length_attr_storage = static_cast<char*>(tuple_slot) + per_tuple_null_bitmap_bytes_;
-
-          std::vector<attribute_id>::const_iterator attr_map_it = attribute_map.begin();
-          for (CatalogRelationSchema::const_iterator attr_it = relation_.begin();
-               attr_it != relation_.end();
-               ++attr_it, ++attr_map_it) {
-            const int nullable_idx = relation_.getNullableAttributeIndex(attr_it->getID());
-            if (nullable_idx != -1) {
-              const void *attr_value = accessor->template getUntypedValue<true>(*attr_map_it);
-              if (attr_value == nullptr) {
-                tuple_null_bitmap.setBit(nullable_idx, true);
-              } else {
-                std::memcpy(fixed_length_attr_storage
-                                + relation_.getFixedLengthAttributeOffset(attr_it->getID()),
-                            attr_value,
-                            attr_it->getType().maximumByteLength());
-              }
+
+        // Find the slot and locate its sub-structures.
+        void *tuple_slot = static_cast<char *>(tuple_storage_) + pos * tuple_slot_bytes_;
+
+        BitVector<true> tuple_null_bitmap(tuple_slot, insertInfo.num_nullable_attrs_);
+        tuple_null_bitmap.clear();
+        char *fixed_length_attr_storage = static_cast<char *>(tuple_slot) + insertInfo.fixed_len_offset_;
+        std::uint32_t *variable_length_info_array =
+          reinterpret_cast<std::uint32_t *>(static_cast<char *>(tuple_slot) + insertInfo.var_len_offset_);
+
+        // Start writing variable-length data at the beginning of the
+        // newly allocated range.
+        std::size_t current_variable_position = tuple_storage_bytes_ - header_->variable_length_bytes_allocated;
+        std::uint32_t current_null_idx = 0;
+        for (attribute_id accessor_attr_id = 0;
+             static_cast<std::size_t >(accessor_attr_id) < insertInfo.num_attrs_; ++accessor_attr_id) {
+          bool nullable = insertInfo.is_nullable_.getBit(accessor_attr_id);
+          bool variable = insertInfo.is_variable_.getBit(accessor_attr_id);
+
+          if (!nullable && !variable) {
+            DCHECK_EQ(-1, relation_.getNullableAttributeIndex(accessor_attr_id));
+
+            const void *attr_value = accessor->template getUntypedValue<false>(attribute_map[accessor_attr_id]);
+            std::memcpy(fixed_length_attr_storage + insertInfo.fixed_len_offsets_[accessor_attr_id],
+                        attr_value,
+                        insertInfo.fixed_len_sizes_[accessor_attr_id]);
+          } else if (nullable && !variable) {
+            DCHECK_EQ(relation_.getNullableAttributeIndex(accessor_attr_id), static_cast<int>(current_null_idx));
+
+            TypedValue attr_value(accessor->getTypedValue(attribute_map[accessor_attr_id]));
```

Comment:
Good catch. Initially I didn't do this because I thought it wouldn't be handled by the Compressed/ValueAccessors. However, scanning their code it seems like this will work, and I will make this change.